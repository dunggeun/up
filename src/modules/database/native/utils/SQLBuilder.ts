/* eslint-disable no-case-declarations */
import dedent from 'dedent';
import { Logger } from 'src/modules/logger';
import type { DatabaseRecord, Value, WhereConditions } from '../../types';

type QueryType = 'select' | 'count' | 'insert' | 'update' | 'delete';

const TAG = 'SQLBuilder';

export class SQLBuilder {
  private _tableName?: string;
  private _type?: QueryType;
  private _values?: DatabaseRecord;
  private _limit?: number;
  private _orderTarget?: string;
  private _by?: 'desc' | 'asc';
  private _condition?: WhereConditions;

  private stringify(value: Value): string {
    if (typeof value === 'string') {
      return `"${value}"`;
    } else if (value === null) {
      return 'null';
    }
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    return `${value}`;
  }

  table(tableName: string): this {
    this._tableName = tableName;
    return this;
  }

  type(type: QueryType): this {
    this._type = type;
    return this;
  }

  values(values: DatabaseRecord): this {
    this._values = values;
    return this;
  }

  where(condition: WhereConditions): this {
    this._condition = condition;
    return this;
  }

  limit(limit: number): this {
    this._limit = limit;
    return this;
  }

  orderBy(target: string, by: 'desc' | 'asc'): this {
    this._orderTarget = target;
    this._by = by;
    return this;
  }

  build(): string {
    if (!(this._type && this._tableName)) {
      throw Error('invalid query type or table name');
    }

    let query: string;
    const WHERE_CONDITION =
      this._condition &&
      Object.entries(this._condition)
        .map(([column, { symbol, value }]) => {
          return `${column} ${symbol} ${this.stringify(value)}`;
        })
        .join(' AND ');
    const WHERE_STATEMENTS = WHERE_CONDITION ? `WHERE ${WHERE_CONDITION}` : '';

    switch (this._type) {
      case 'select':
        const shouldOrdering = this._orderTarget && this._by;
        const ORDER_BY_STATEMENTS = shouldOrdering
          ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            `ORDER BY ${this._orderTarget!} ${this._by!}`
          : '';

        const LIMIT_CONSTRAINT =
          this._limit !== undefined ? `LIMIT ${this._limit}` : '';

        query = dedent`
          SELECT * FROM ${this._tableName}
          ${WHERE_STATEMENTS}
          ${ORDER_BY_STATEMENTS}
          ${LIMIT_CONSTRAINT};
        `;
        break;

      case 'count':
        query = dedent`SELECT count(*) as count FROM ${this._tableName};`;
        break;

      case 'insert':
        const VALUE_COLUMNS = Object.keys(this._values ?? {}).join(',');
        const VALUES = Object.values(this._values ?? {})
          .map((value) => this.stringify(value))
          .join(',');

        query = dedent`
          INSERT INTO ${this._tableName} (${VALUE_COLUMNS})
          VALUES (${VALUES});
        `;
        break;

      case 'update':
        const SET_STATEMENTS = Object.entries(this._values ?? {})
          .map(([column, value]) => `${column} = ${this.stringify(value)}`)
          .join(',');

        query = dedent`
          UPDATE ${this._tableName}
          SET ${SET_STATEMENTS}
          ${WHERE_STATEMENTS};
        `;
        break;

      case 'delete':
        query = dedent`
          DELETE FROM ${this._tableName}
          ${WHERE_STATEMENTS};
        `;
        break;
    }

    query = query.replace(/[\r\n]/g, ' ');
    Logger.debug(TAG, query);

    return query;
  }
}
