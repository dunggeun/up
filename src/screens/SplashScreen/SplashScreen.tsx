import React from 'react';
import { StyleSheet, Text } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',
  },
  text: {
    color: 'black',
  },
});

export function SplashScreen(): JSX.Element {
  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.text}>Splash</Text>
    </SafeAreaView>
  );
}
