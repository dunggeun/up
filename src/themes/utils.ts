import { Platform } from 'react-native';

export const webFont = (font: string): string => {
  return Platform.select({
    web: `${font}, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, Inter-serif`,
    default: font,
  });
};
