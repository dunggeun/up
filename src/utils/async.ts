export const delay = (duration: number): Promise<void> => {
  return new Promise((resolve) => {
    setTimeout(resolve, duration);
  });
};

/* istanbul ignore next */

export const debounce = <T>(
  callback: (...args: T[]) => void,
  // eslint-disable-next-line @typescript-eslint/no-shadow
  delay: number,
): ((...args: T[]) => void) => {
  let timer: NodeJS.Timeout;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => callback(...args), delay);
  };
};
