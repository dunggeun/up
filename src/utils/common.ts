import { Platform } from 'react-native';
import {
  trigger,
  HapticFeedbackTypes as RNHapticFeedbackTypes,
} from 'react-native-haptic-feedback';

const HAPTIC_FEEDBACK_TYPES = {
  press: Platform.select<RNHapticFeedbackTypes>({
    ios: RNHapticFeedbackTypes.impactLight,
    android: RNHapticFeedbackTypes.virtualKey,
  }),
  buttonPress: Platform.select<RNHapticFeedbackTypes>({
    ios: RNHapticFeedbackTypes.impactMedium,
    android: RNHapticFeedbackTypes.effectClick,
  }),
  buttonLongPress: Platform.select<RNHapticFeedbackTypes>({
    ios: RNHapticFeedbackTypes.impactHeavy,
    android: RNHapticFeedbackTypes.effectHeavyClick,
  }),
} as const;

export type HapticFeedbackTypes = keyof typeof HAPTIC_FEEDBACK_TYPES;

/* istanbul ignore next */
// eslint-disable-next-line @typescript-eslint/no-empty-function
export function noop(): void {}

export function triggerHaptic(type: HapticFeedbackTypes = 'press'): void {
  const feedbackType = HAPTIC_FEEDBACK_TYPES[type];
  feedbackType && trigger(feedbackType);
}
