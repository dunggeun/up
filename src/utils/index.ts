export * from './async';
export * from './common';
export * from './locales';
