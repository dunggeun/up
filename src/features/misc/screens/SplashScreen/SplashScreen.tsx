import React from 'react';
import { styled, View, Image } from 'dripsy';
import Logo from 'src/assets/images/logo.png';
import {
  APP_BAR_HEIGHT,
  LANDING_LOGO_SIZE,
  LANDING_LOGO_MARGIN,
} from 'src/constants';
import { CommonLayout } from 'src/designs';

const LogoArea = styled(View)({
  flex: 1,
  justifyContent: 'center',
  alignItems: 'center',
  paddingTop: APP_BAR_HEIGHT,
  paddingBottom: LANDING_LOGO_MARGIN,
});

const LogoImage = styled(Image)({
  width: LANDING_LOGO_SIZE,
  height: LANDING_LOGO_SIZE,
});

export function SplashScreen(): JSX.Element {
  return (
    <CommonLayout>
      <LogoArea>
        <LogoImage source={Logo} />
      </LogoArea>
    </CommonLayout>
  );
}
