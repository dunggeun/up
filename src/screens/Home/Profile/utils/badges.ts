import { t } from 'src/translations';
import dummy from 'src/assets/badges/dummy.png';

import type { ImageSourcePropType } from 'react-native';

const BADGE_SET = [[t('badge.0'), dummy]] as const;
const FALLBACK_BADGE = ['', dummy] as const;

export const getBadge = (id: number): { title: string, image: ImageSourcePropType } => {
  const badge = BADGE_SET[id];
  const [title, image] = badge ?? FALLBACK_BADGE;
  return { title, image };
};
