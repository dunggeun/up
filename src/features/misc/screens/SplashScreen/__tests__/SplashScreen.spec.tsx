import React from 'react';
import { render, cleanup } from '@testing-library/react-native';
import { withDripsy } from 'tests/utils';
import { SplashScreen } from '../SplashScreen';

describe('screens/SplashScreen', () => {
  afterEach(() => {
    cleanup();
  });

  describe('렌더링 되었을 때', () => {
    it('스냅샷이 일치해야 한다', () => {
      const tree = render(withDripsy(<SplashScreen />)).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
