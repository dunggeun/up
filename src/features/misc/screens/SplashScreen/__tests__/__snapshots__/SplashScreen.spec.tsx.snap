// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screens/SplashScreen 렌더링 되었을 때 스냅샷이 일치해야 한다 1`] = `
<View
  style={
    [
      {},
      {
        "flex": 1,
      },
      undefined,
      {
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
      },
      {},
      {},
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          undefined,
          {
            "flex": 1,
          },
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignSelf": "center",
            "backgroundColor": "#ffffff",
            "flex": 1,
            "marginHorizontal": "auto",
            "maxWidth": 600,
            "paddingHorizontal": 16,
            "width": "100%",
          },
          undefined,
          {},
        ]
      }
    >
      <View
        style={
          [
            {},
            {
              "flex": 1,
            },
            {},
          ]
        }
      >
        <View
          style={
            [
              {},
              {
                "flex": 1,
                "gap": 16,
              },
              undefined,
              {},
              {},
            ]
          }
        >
          <View
            style={
              [
                {},
                {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
                {},
                {},
              ]
            }
          >
            <Image
              source={
                {
                  "testUri": "../../../src/assets/images/logo.png",
                }
              }
              style={
                [
                  {},
                  {
                    "height": 180,
                    "width": 180,
                  },
                  undefined,
                  {},
                  {},
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
